core/client/src/main/java/alluxio/client/file/FileInStream.java

```diff
@@ -528,20 +530,26 @@ private void updateCacheStream(long blockId) throws IOException {
       return;
     }
 
-    // If this block is read from a remote worker but we don't have a local worker, don't cache
-    if (isReadingFromRemoteBlockWorker() && !mContext.hasLocalWorker()) {
-      return;
-    }
-
     // Unlike updateBlockInStream below, we never start a block cache stream if mPos is in the
     // middle of a block.
     if (mPos % mBlockSize != 0) {
       return;
     }
 
     try {
-      WorkerNetAddress address = mLocationPolicy.getWorkerForNextBlock(
-          mBlockStore.getWorkerInfoList(), getBlockSizeAllocation(mPos));
+      // If this block is read from a remote worker, we should never cache except to a local worker.
+      if (isReadingFromLocalBlockWorker()) {
+        WorkerNetAddress localWorker = mContext.getLocalWorker();
+        if (localWorker != null) {
+          mCurrentCacheStream =
+              mBlockStore.getOutStream(blockId, getBlockSize(mPos), localWorker, mOutStreamOptions);
```

Comment:
this overrides the location policy. It looks like you always cache to local if reading from remote.

do you think it is easier if we do one of the following?
1) Do not cache if reading from remote unless policy tells us to cache to local? I think this is what the origin code ('if (isReadingFromRemoteBlockWorker() && !mContext.hasLocalWorker())') is trying to do.
2) Get the address that we are reading from RemoteBlockInStream#location(). Compare that with the host to cache here. If equal, skip caching.